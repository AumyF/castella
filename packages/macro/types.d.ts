import React from "react";

declare const slotBland: unique symbol;

export type Slot<SlotName extends string> = {
  [slotBland]: SlotName;
};

/**
 * Converts given CSS string for use in Castella.
 */
export function css(
  arr: TemplateStringsArray,
  ...args: readonly unknown[]
): string;

export type HtmlTemplateItem<SlotName extends string> = string | Slot<SlotName>;

export type HtmlResult<SlotName extends string> = {
  shadowHtml: string;
  slots: readonly SlotName[];
};

/**
 * Converts given HTML string for use in Castella.
 */
export function html<SlotName extends string>(
  arr: TemplateStringsArray,
  ...args: readonly HtmlTemplateItem<SlotName>[]
): HtmlResult<Exclude<SlotName, "children" | "">>;

export type CastellaComponent<
  SlotName extends string
> = React.FunctionComponent<
  string extends SlotName
    ? {}
    : {
        [K in SlotName]?: React.ReactNode;
      }
>;

/**
 * Creates a slot.
 */
export function slot(): Slot<"">;
export function slot<SlotName extends string>(name: SlotName): Slot<SlotName>;

type ShadowAllowElementName =
  | "article"
  | "aside"
  | "blockquote"
  | "body"
  | "div"
  | "footer"
  | "h1"
  | "h2"
  | "h3"
  | "h4"
  | "h5"
  | "h6"
  | "header"
  | "main"
  | "nav"
  | "p"
  | "section"
  | "span";

type CastellaFunction = <SlotName extends string>(
  css: string,
  html: HtmlResult<SlotName>
) => CastellaComponent<SlotName>;

// export const castella: CastellaFunction &
//   Record<ShadowAllowElementName, CastellaFunction>;

/**
 * Creates a React component from given CSS and HTML.
 */
export const castella: CastellaFunction;

export const styled: {
  /**
   * Creates a React component with given CSS attached.
   */
  [Elm in keyof JSX.IntrinsicElements]: (
    arr: TemplateStringsArray,
    ...values: readonly string[]
  ) => React.FunctionComponent<JSX.IntrinsicElements[Elm]>;
};
