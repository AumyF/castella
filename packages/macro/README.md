# @castella/macro

Visit [the repository for Castella](https://github.com/uhyo/castella) for details.