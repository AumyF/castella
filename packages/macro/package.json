{
  "name": "@castella/macro",
  "version": "1.0.0",
  "description": "CSS in JS for React",
  "main": "lib/index.js",
  "types": "./types.d.ts",
  "scripts": {
    "build": "tsc",
    "watch": "tsc --watch",
    "test": "jest",
    "test:update": "jest -u",
    "prepublishOnly": "npm run build && npm run test"
  },
  "keywords": [
    "babel-plugin-macros"
  ],
  "author": "uhyo <uhyo@uhy.ooo>",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/uhyo/castella.git"
  },
  "files": [
    "lib/**/*",
    "src/**/*",
    "types.d.ts"
  ],
  "dependencies": {
    "babel-plugin-macros": "^2.8.0",
    "stylis": "^4.0.3"
  },
  "devDependencies": {
    "@babel/core": "^7.12.3",
    "@babel/types": "^7.12.1",
    "@types/babel-plugin-macros": "^2.8.3",
    "@types/babel-types": "^7.0.9",
    "@types/node": "^14.11.10",
    "@types/react": "^16.9.53",
    "@types/stylis": "^4.0.0",
    "babel-plugin-tester": "^10.0.0",
    "jest": "^26.5.3",
    "typescript": "^4.0.3"
  },
  "peerDependencies": {
    "@types/react": "^16.9.53"
  }
}