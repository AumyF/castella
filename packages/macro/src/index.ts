import { createMacro } from "babel-plugin-macros";

// `source` is not in @types/babel-plugin-macros :(
// @ts-expect-error
export = createMacro(({ references, state, babel, source }) => {
  const cssFuncReferences = [...(references.css || [])];
  const htmlFuncReferences = [...(references.html || [])];
  const wc = [...(references.wc || [])];
});
