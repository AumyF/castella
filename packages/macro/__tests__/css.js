const pluginTester = require("babel-plugin-tester").default;
const plugin = require("babel-plugin-macros");

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: {
    placeholders: `
      import { castella, css, html, slot } from '../../macro'

      const A = castella(
        css\`
          display: flex;

          margin: calc(1em + \${"3px"});
          color: \${color};
        \`,
        html\`
          <div>\${slot()}</div>
        \`
      );
    `,
    nested: `
      import { castella, css, html, slot } from '../../macro'

      const A = castella(
        css\`
          display: flex;

          .left {
            flex: 100px 0 0;

            font-size: \${size}
          }

          & > .right {
            flex: auto 1 0;
          }
        \`,
        html\`
          <div>\${slot()}</div>
        \`
      );
    `,
    cannotConvert: {
      code: `
        import { castella, css, html, slot } from '../../macro'

        const A = castella(
          css\`
            display: flex;

            \${"padding: 10px"};
          \`,
          html\`
            <div>\${slot()}</div>
          \`
        );
      `,
      snapshot: false,
      error: "Could not convert CSS template."
    }
  },
});

