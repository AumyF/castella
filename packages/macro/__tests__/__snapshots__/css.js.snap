// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin nested: nested 1`] = `

import { wc, css, html, slot } from '../../macro'

const A = wc(
  css\`
    display: flex;

    .left {
      flex: 100px 0 0;

      font-size: \${size}
    }

    & > .right {
      flex: auto 1 0;
    }
  \`,
  html\`
    <div>\${slot()}</div>
  \`
);

      ↓ ↓ ↓ ↓ ↓ ↓

const A = wc(
  \`:host{display:flex;}.left{flex:100px 0 0;font-size:\${size};}:host>.right{flex:auto 1 0;}\`,
  {
    html: \`<div><slot></slot></div>\`,
  }
);


`;

exports[`unknown plugin placeholders: placeholders 1`] = `

import { wc, css, html, slot } from '../../macro'

const A = wc(
  css\`
    display: flex;

    margin: calc(1em + \${"3px"});
    color: \${color};
  \`,
  html\`
    <div>\${slot()}</div>
  \`
);

      ↓ ↓ ↓ ↓ ↓ ↓

const A = wc(
  \`:host{display:flex;margin:calc(1em + \${"3px"});color:\${color};}\`,
  {
    html: \`<div><slot></slot></div>\`,
  }
);


`;
